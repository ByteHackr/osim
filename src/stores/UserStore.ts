import {ref, computed} from 'vue';
import {defineStore} from 'pinia';
import {z} from 'zod';
import jwtDecode from 'jwt-decode';
import type {JwtPayload} from 'jwt-decode';

import router from '@/router';
import {osimRuntime} from '@/stores/osimRuntime';

const _sessionStorageKey = 'UserStore';

const loginResponse = z.object({
  access: z.string(),
  refresh: z.string(),
  env: z.string(),
});

export const useUserStore = defineStore('UserStore', () => {


  const refresh = ref<string>('');

  const jwtRefresh = computed<JwtPayload | null>(() => {
    try {
      return jwtDecode(refresh.value);
    } catch (e) {
      console.debug('UserStore: refresh token not a valid JWT', refresh.value, e);
    }
    return null;
  });

  const userName = computed(() => {
    if (refresh.value === '') {
      return 'Not Logged In';
    }
    let sub = jwtRefresh.value?.sub;
    if (sub == null) {
      // @ts-ignore
      sub = 'User ID: ' + jwtAccess.value?.user_id;
    }
    if (sub == null) {
      // fallback: look for user in domain root
      sub = document.cookie
          .split(';')
          .map(x => x.trim())
          .filter(x => /^.._user=/.test(x))
          .map(x => x.substring('.._user='.length))
          .reduce((acc: string[], x) => acc.concat(x.split('|')), [])
          .map(x => decodeURIComponent(x))
          .find(x => true)
    }
    if (sub == null) {
      sub = 'Current User';
    }
    return sub;
  });

  const env = ref<string>(''); // Placeholder: environment

  // let refreshInterval = 30000;
  // let refreshId = 0;


  function setTokens(refresh_: string, env_: string) {
    refresh.value = refresh_;
    env.value = env_;
    // refreshId = setInterval(performTokenRefresh, 5000);
    // writeSessionStorage()
  }

  readSessionStorage(); // Initially read when loading UserStore
  function readSessionStorage() {
    let storedUserStore = sessionStorage.getItem(_sessionStorageKey);
    if (storedUserStore != null) {
      try {
        storedUserStore = JSON.parse(storedUserStore);
        // @ts-ignore
        access.value = storedUserStore.access;
        // @ts-ignore
        refresh.value = storedUserStore.refresh;
        // @ts-ignore
        env.value = storedUserStore.env;
      } catch (e) {
        console.error('UserStore: unable to restore from sessionStorage', e);
        $reset();
      }
    }
  }

  // function login(jwtAccess, jwtRefresh) {
  //   return userService.login()
  //       .then(user => {
  //         jwt.value.user = JSON.stringify(user);
  //       })
  // }
  async function login() {
    // return fetch('https://osidb-stage.prodsec.redhat.com/auth/token', {
    return fetch(`${osimRuntime.value.backends.osidb}/auth/token`, {
      // credentials: 'same-origin',
      credentials: 'include',
      cache: 'no-cache',
    })
        .then(response => response.json())
        .then(json => {
          const parsedLoginResponse = loginResponse.passthrough().parse(json);
          access.value = parsedLoginResponse.access;
          refresh.value = parsedLoginResponse.refresh;
          env.value = parsedLoginResponse.env;
        })
        .catch(e => {
          console.error('UserStore: unsuccessful login request', e);
        });
  }

  // function writeSessionStorage() {
  //   let storedUserStore = {
  //     access: access.value,
  //     refresh: refresh.value,
  //     env: _env.value,
  //     _modifyDate: Date.now(),
  //   };
  //   sessionStorage.setItem(_sessionStorageKey, JSON.stringify(storedUserStore));
  // }

  function logout() {
    // clearInterval(refreshInterval);
    $reset();

    // router.push({name: 'login'});
    // return Promise.resolve()
    //     .then(() => {
    //       jwt.value.user = undefined;
    //     })
    //     .then(() => {
    //       router.push({name: 'login'})
    //     })
    return router.push({name: 'login'});
  }

  /**
   * Side effect: wipes tokens if tokens are expired
   */
  function isAuthenticated() {
    let now = Date.now();
    let refreshExp = jwtRefresh.value?.exp;
    if (refreshExp != null) {
      return now < refreshExp * 1000;
    }
    $reset();
    return false;
  }

  function $reset() {
    setTokens('', '');
  }

  return {
    refresh,
    jwtRefresh,
    userName,
    env,
    login,
    logout,
    isAuthenticated,
    $reset,
  };
});

