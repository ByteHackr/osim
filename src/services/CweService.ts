import { osimRuntime } from '@/stores/osimRuntime';
import type { CWEMemberType } from '@/types/mitreCwe';

export const DATA_KEY = 'CWE:API-DATA';
const VERSION_KEY = 'CWE:API-VERSION';

interface CweViews {
  Views: {
    Members: {
      CweID: string;
    }[];
  }[];
}

interface CweCategories {
  Categories: {
    ID: string;
    MappingNotes: {
      Usage: string;
    };
    Name: string;
    Relationships: Array<any>;
    Status: string;
    Summary: string;
  }[];
}

interface CweWeaknesses {
  Weaknesses: {
    Description: string;
    ID: string;
    MappingNotes: {
      Usage: string;
    };
    Name: string;
    Status: string;
  }[];
};

export function loadCweData(): CWEMemberType[] {
  const data = localStorage.getItem(DATA_KEY);
  let cweData: CWEMemberType[] = [];

  try {
    cweData = JSON.parse(data || '[]');
  } catch (e) {
    console.error('CweService:loadCweData() Failed to parse CWE data:', e);
  }

  return cweData;
};

export async function updateCWEData() {
  const baseUrl = osimRuntime.value.backends.mitre;
  try {
    if (!baseUrl) {
      console.debug('No Mitre backed configured, skipping CWE API cache update.');
      return;
    }

    const [version, isNew] = await checkNewVersion(baseUrl);
    if (!isNew) {
      console.debug('✅ CWE API cache is up-to-date.');
      return;
    }

    fetchAndCache(baseUrl);

    localStorage.setItem(VERSION_KEY, version);
  } catch (error) {
    console.error('CweService::fetchAndCacheAPI() Error getting CWE data', error);
  }
}

async function checkNewVersion(baseUrl: string): Promise<[string, boolean]> {
  const version = await fetch(`${baseUrl}/cwe/version`);

  if (!version.ok) {
    throw new Error('Failed to fetch CWE version data');
  }

  const versionData = await version.json();
  const storedVersion = localStorage.getItem(VERSION_KEY);

  const isNew = storedVersion !== versionData.ContentVersion;
  return [versionData.ContentVersion, isNew];
}

async function fetchAndCache(baseUrl: string) {
  const cweView = await fetchCweView(baseUrl);
  const cweCategories = await fetchCweCategories(baseUrl, cweView);
  const cweWeaknesses = await fetchCweWeaknesses(baseUrl, cweCategories);
  localStorage.setItem(DATA_KEY, JSON.stringify(cweWeaknesses));
  console.debug('✅ CWE API cache updated.');
}

async function fetchCweView(baseUrl: string) {
  // 699 is the id for the "Software Development" CWE view
  const view = await fetch(`${baseUrl}/cwe/view/699`);
  if (!view.ok) {
    throw new Error('Failed to fetch CWE OpenAPI data');
  }
  const cweData: CweViews = await view.json();
  const cweCategoryIds = cweData.Views[0].Members.map(member => member.CweID);
  return cweCategoryIds;
}

async function fetchCweCategories(baseUrl: string, cweCategoryIds: string[]) {
  const response = await fetch(`${baseUrl}/cwe/category/${cweCategoryIds.join(',')}`);
  if (!response.ok) {
    throw new Error('Failed to fetch CWE categories data');
  }

  const cweCategoryData: CweCategories = await response.json();
  const cweWeaknessIds = cweCategoryData.Categories.flatMap(member =>
    member.Relationships.map(relationship => relationship.CweID),
  );
  return cweWeaknessIds;
}

async function fetchCweWeaknesses(baseUrl: string, cweIds: string[]) {
  const response = await fetch(`${baseUrl}/cwe/weakness/${cweIds.join(',')}`);
  if (!response.ok) {
    throw new Error('Failed to fetch detailed CWE data');
  }

  const cweWeaknessesData: CweWeaknesses = await response.json();
  const cweWeaknesses: CWEMemberType[] = cweWeaknessesData.Weaknesses.map(weakness => ({
    id: weakness.ID,
    name: weakness.Name,
    status: weakness.Status,
    summary: weakness.Description,
    usage: weakness.MappingNotes.Usage,
  }));
  return cweWeaknesses;
}
