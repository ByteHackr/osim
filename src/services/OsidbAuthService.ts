import axios from 'axios';
import {osimRuntime} from '@/stores/osimRuntime';
import {useUserStore} from '../stores/UserStore';
import {z} from 'zod';
import type {AxiosRequestConfig} from 'axios';

const RefreshResponse = z.object({
  access: z.string()
});


export async function osidbFetch(config: AxiosRequestConfig) {
  try {
    const accessToken = await getNextAccessToken();
    config.headers = config.headers || {};
    config.headers['Authorization'] = `Bearer ${accessToken}`;
    config.baseURL = osimRuntime.value.backends.osidb;
    config.withCredentials = true;
    return axios(config);
  } catch (e) {
    throw e;
  }
}

export async function getNextAccessToken() {
  const userStore = useUserStore(); // Moving this to module scope results in "cannot access before initialization" - probably a vite or typescript bug
  let response;
  try {
    response = await axios({
      method: 'post',
      url: '/auth/token/refresh',
      baseURL: osimRuntime.value.backends.osidb,
      withCredentials: true,
      data: {
        refresh: userStore.refresh,
      },
      // headers: {
      //   'Authorization': `Bearer`
      // },
    });
  } catch (e) {
    return userStore.logout()
        .finally(() => {
          throw new Error('Refresh token expired');
        })
  }
  try {
    const responseBody = response.data;
    const parsedBody = RefreshResponse.parse(responseBody);
    return parsedBody.access;
  } catch (e) {
    return userStore.logout()
        .finally(() => {
          throw new Error('Unable to parse next access token');
        })
  }
}
