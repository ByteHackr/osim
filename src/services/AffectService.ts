import { osidbFetch } from '@/services/OsidbAuthService';
import { createCatchHandler, createSuccessHandler } from '@/composables/service-helpers';

// export async function putAffect(uuid: string, affectObject: ZodAffectType) {
export async function putAffect(uuid: string, affectObject: any) {
  return osidbFetch({
    method: 'put',
    url: `/osidb/api/v1/affects/${uuid}`,
    data: affectObject,
  }).then((response) => {
    console.log(response);
    return response.data;
  });
}

// export async function postAffect(affectObject: ZodAffectType) {
export async function postAffect(affectObject: any) {
  return osidbFetch({
    method: 'post',
    url: `/osidb/api/v1/affects`,
    data: affectObject,
  }).then((response) => {
    console.log(response);
    return response.data;
  });
}

export async function deleteAffect(uuid: string) {
  return osidbFetch({
    method: 'delete',
    url: `/osidb/api/v1/affects/${uuid}`,
  })
    .then(createSuccessHandler({ title: 'Success!', body: 'Affect Deleted.' }))
    .catch(createCatchHandler('Error deleting Affect:'));
}

type AffectAcknowledgmentPost = {
  name: string;
  affiliation: string;
  from_upstream: boolean;
  embargoed: boolean;
};

export async function postAffectAcknowledgment(
  flawId: string,
  requestBody: AffectAcknowledgmentPost
) {
  return osidbFetch({
    method: 'post',
    url: `/osidb/api/v1/flaws/${flawId}/acknowledgments`,
    data: requestBody,
  })
    .then(createSuccessHandler({ title: 'Success!', body: 'Acknowledgment created.' }))
    .catch(createCatchHandler('Error creating Acknowledgment:'));
}

type AffectAcknowledgmentPut = AffectAcknowledgmentPost & {
  updated_dt: string;
};

export async function putAffectAcknowledgment(
  flawId: string,
  acknowledgmentId: string,
  requestBody: AffectAcknowledgmentPut
) {
  return osidbFetch({
    method: 'put',
    url: `/osidb/api/v1/flaws/${flawId}/acknowledgments/${acknowledgmentId}`,
    data: requestBody,
  })
    .then(createSuccessHandler({ title: 'Success!', body: 'Acknowledgment updated.' }))
    .catch(createCatchHandler('Error updating Acknowledgment:'));
}

