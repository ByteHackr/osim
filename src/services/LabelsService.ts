import { createCatchHandler, createSuccessHandler } from '@/composables/service-helpers';

import type { PaginatedFlawLabelList } from '@/generated-client';
import type { ZodFlawLabelType } from '@/types/zodFlaw';

import { osidbFetch } from './OsidbAuthService';

export async function fetchLabels() {
  try {
    const { data }: { data: PaginatedFlawLabelList } = await osidbFetch({
      method: 'get',
      url: '/osidb/api/v1/labels',
    });

    return data.results;
  } catch (error) {
    console.error('LabelService::fetchLabels() Error fetching labels', error);
    return [];
  }
}

export async function createLabel(flawUUID: string, label: ZodFlawLabelType) {
  return osidbFetch({
    method: 'post',
    url: `/osidb/api/v1/flaws/${flawUUID}/labels`,
    data: label,
  })
    .then(createSuccessHandler({ title: 'Success!', body: `Label ${label.label} created.` }))
    .catch(createCatchHandler(`Error creating label ${label.label}`));
}

export async function deleteLabel(flawUUID: string, label: ZodFlawLabelType) {
  return osidbFetch({
    method: 'delete',
    url: `/osidb/api/v1/flaws/${flawUUID}/labels/${label.uuid}`,
  })
    .then(createSuccessHandler({ title: 'Success!', body: `Label ${label.label} deleted.` }))
    .catch(createCatchHandler(`Error deleting label ${label.label}`));
}

export async function updateLabel(flawUUID: string, label: ZodFlawLabelType) {
  return osidbFetch({
    method: 'put',
    url: `/osidb/api/v1/flaws/${flawUUID}/labels/${label.uuid}`,
    data: label,
  })
    .then(createSuccessHandler({ title: 'Success!', body: `Label ${label.label} updated.` }))
    .catch(createCatchHandler(`Error updating label ${label.label}`));
}
