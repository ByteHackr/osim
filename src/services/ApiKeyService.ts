//    <div class="alert alert-danger"
//          role="alert"
//         v-if="!/.+/.test(settingsStore.settings.bugzillaApiKey ?? '')">
//       You have not set your Bugzilla API key in this tab!<br/>
//       Flaw creation requires your Bugzilla API key to be set.<br/>
//       Visit <RouterLink :to="{name: 'settings'}">Settings</RouterLink>.
//     </div>

import { createCatchHandler } from '@/composables/service-helpers';

import { useSettingsStore } from '@/stores/SettingsStore';
import { useToastStore } from '@/stores/ToastStore';
import { osidbFetch } from '@/services/OsidbAuthService';
// import { useRouter } from 'vue-router';

export interface IntegrationTokensResponse {
  bugzilla: null | string;
  jira: null | string;
}

export interface IntegrationTokensPatchRequest {
  bugzilla?: string;
  jira?: string;
}

export function notifyApiKeyUnset() {
  const settingsStore = useSettingsStore();
  const { bugzillaApiKey, jiraApiKey } = settingsStore.apiKeys;
  const unsetKeys: string[] = [];

  if (!bugzillaApiKey) {
    unsetKeys.push('Bugzilla');
  }

  if (!jiraApiKey) {
    unsetKeys.push('JIRA');
  }

  if (!bugzillaApiKey || !jiraApiKey) {
    const { addToast } = useToastStore();
    const lis = unsetKeys.map(key => `<li>${key}</li>`).join('');

    addToast({
      css: 'warning',
      title: 'Cannot Edit/Create Flaws',
      bodyHtml: true,
      body: `You have not set the following keys in this tab: <ul>${lis}</ul>` +
      'Flaw creation requires your Bugzilla API or JIRA key to be set. ' +
      'Visit Settings to set any required keys.',
    });
    // maybeRedirectToSettings();
  }
}

/**
 * Save API keys to the backend
 */
export async function saveApiKeysToBackend(apiKeys: IntegrationTokensPatchRequest) {
  console.log('ApiKeyService: saveApiKeysToBackend called with:', apiKeys);

  return osidbFetch({
    method: 'PATCH',
    url: '/osidb/integrations',
    data: apiKeys,
  })
    .catch(createCatchHandler('Failed to save API keys'));
}

/**
 * Retrieve API keys from the backend
 */
export async function getApiKeysFromBackend(): Promise<IntegrationTokensResponse> {
  try {
    const response = await osidbFetch({
      method: 'get',
      url: '/osidb/integrations',
    });
    return response.data as IntegrationTokensResponse;
  } catch (error) {
    console.error('Failed to retrieve API keys:', error);
    throw error;
  }
}

// function maybeRedirectToSettings () {
//   const router = useRouter();
//   if (router.currentRoute.value.fullPath === '/flaws/new') {
//     router.push({path: '/settings'});
//   }
// }
