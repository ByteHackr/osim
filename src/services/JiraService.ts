import { createSuccessHandler, createCatchHandler } from '@/composables/service-helpers';

import { useSettingsStore } from '@/stores/SettingsStore';
import { useToastStore } from '@/stores/ToastStore';
import { getNextAccessToken } from '@/services/OsidbAuthService';
import {
  osimRuntime,
} from '@/stores/osimRuntime';
import type { ZodJiraUserAssignableType, ZodJiraIssueType } from '@/types/zodJira';

type JiraFetchCallbacks = {
  beforeFetch?: (options: JiraFetchOptions) => Promise<void> | void;
};

type JiraGetFetchOptions = {
  data?: never;
  method: 'GET' | 'get';
  params?: Record<string, any>;
  url: string;
};

type JiraPutPostFetchOptions = {
  data?: Record<string, any>;
  method: 'POST' | 'post' | 'PUT' | 'put';
  params?: Record<string, any>;
  url: string;
};

type JiraFetchOptions =
  | JiraGetFetchOptions
  | JiraPutPostFetchOptions;

async function jiraFetch<T = any>(config: JiraFetchOptions, factoryOptions?: JiraFetchCallbacks) {
  if (factoryOptions?.beforeFetch) {
    await factoryOptions.beforeFetch(config);
  }

  const body = config?.data ? JSON.stringify(config.data) : undefined;
  const queryString = queryStringFromParams(config?.params ?? {});
  const baseUrl = osimRuntime.value.backends.jira;

  if (config?.method?.toUpperCase() !== 'GET' && osimRuntime.value.readOnly) {
    useToastStore().addToast({ title: 'Operation Not Permitted', body: 'OSIM is in read-only mode.' });
    return Promise.reject('OSIM is in read-only mode.');
  }

  let response: Response;
  try {
    response = await fetch(`${baseUrl}${config.url}${queryString}`, {
      method: config?.method ?? 'GET',
      headers: await osimRequestHeaders(),
      mode: 'cors',
      credentials: 'include',
      body,
    });
  } catch (e) {
    return Promise.reject(e);
  }

  if (!response.ok) {
    let data;
    if (response.headers.get('content-type')?.startsWith('text/html')) {
      data = await response.text();
    } else {
      data = await response.json();
    }

    return Promise.reject({
      response: {
        headers: Object.fromEntries(response.headers.entries()),
        ok: response.ok,
        redirected: response.redirected,
        status: response.status,
        statusText: response.statusText,
        type: response.type,
        url: response.url,
        data,
      },
      data,
    },
    );
  }

  if (response.status === 204) {
    return { data: undefined, response };
  }
  return { data: await response.json() as T, response };
}

export async function getJiraComments(taskId: string) {
  return jiraFetch({
    method: 'get',
    url: `/rest/api/2/issue/${taskId}/comment`,
  });
}

export async function searchJiraUsers(query: string, issueKey: string) {
  return jiraFetch<{ users: ZodJiraUserAssignableType[] }>({
    method: 'get',
    url: '/rest/internal/2/users/assignee',
    params: { issueKey, query },
  });
}

export async function getJiraIssue(taskId: string) {
  return jiraFetch<ZodJiraIssueType>({
    method: 'get',
    url: `/rest/api/2/issue/${taskId}`,
  });
}

type putIssueOptions<T, U> = {
  data: {
    fields?: keyof T extends keyof U ? never : T;
    update?: keyof U extends keyof T ? never : U;
  };
  params?: {
    notifyUsers?: boolean;
    returnIssue?: boolean;
  };
};

type putIssueUpdate = Record<string, Array<{
  [_ in 'add' | 'copy' | 'edit' | 'remove' | 'set']?: any;
}>>;

export async function putJiraIssue<T extends object | undefined,
  U extends putIssueUpdate | undefined>(taskId: string, opts: putIssueOptions<T, U>) {
  return jiraFetch({
    method: 'put',
    url: `/rest/api/2/issue/${taskId}`,
    ...opts,
  });
}

export async function postJiraComment(taskId: string, comment: string) {
  return jiraFetch({
    method: 'post',
    url: `/rest/api/2/issue/${taskId}/comment`,
    data: {
      body: comment,
    },
  })
    .then(createSuccessHandler({ title: 'Success!', body: 'Internal comment saved' }))
    .then(response => response.data)
    .catch(createCatchHandler('Error saving internal Jira comment'));
}

export async function getJiraUsername() {
  return jiraFetch({
    method: 'get',
    url: '/rest/api/2/myself',
  }).then((res) => {
    if (res.response.ok) {
      return res.data.name;
    }
  }).catch((error) => {
    console.error('JiraService::getJiraUsername() Error getting your username from jira', error);
    useToastStore().addToast({
      title: 'Jira Error',
      body: `Error getting your username from <a href="${osimRuntime.value.backends.jiraDisplay}">Jira</a>. ` +
      'Is your Jira Access Token valid?',
      bodyHtml: true,
      css: 'warning',
    });
    return '';
  });
}

async function osimRequestHeaders() {
  const settingsStore = useSettingsStore();
  const osidbToken = await getNextAccessToken();
  if (osimRuntime.value.env === 'prod') {
    return new Headers({
      'Authorization': `Bearer ${settingsStore.settings.jiraApiKey}`,
      'Content-Type': 'application/json',
    });
  } else {
    return new Headers({
      'Authorization': `Bearer ${osidbToken}`,
      'Content-Type': 'application/json',
      'Bugzilla-Api-Key': settingsStore.settings.bugzillaApiKey,
      'Jira-Api-Key': settingsStore.settings.jiraApiKey,
      'User-Agent': 'OSIM',
    });
  }
}

function paramsFrom(params: Record<string, any>) {
  const urlParams = new URLSearchParams();
  for (const key in params) {
    if (params[key] !== undefined) {
      urlParams.set(key, params[key]);
    }
  }
  return urlParams;
}

function queryStringFromParams(params: Record<string, any>) {
  const urlParams = paramsFrom(params);
  const queryString = urlParams.toString();
  return queryString ? `?${queryString}` : '';
}

export function jiraTaskUrl(id: string): string {
  return (new URL(`/browse/${id}`, osimRuntime.value.backends.jiraDisplay)).href;
}

export function jiraUserUrl(name: string): string {
  return (new URL(`/ViewProfile.jspa?name=${name}`, osimRuntime.value.backends.jiraDisplay)).href;
}
