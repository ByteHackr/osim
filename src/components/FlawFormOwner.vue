<script setup lang="ts">
import { nextTick, computed, ref, toValue } from 'vue';
import sanitize from 'sanitize-html';
import { useMemoize } from '@vueuse/core';
import { createCatchHandler } from '@/composables/service-helpers';
import { searchJiraUsers } from '@/services/JiraService';
import { useUserStore } from '@/stores/UserStore';
import type { ZodJiraUserPickerType } from '@/types/zodJira';
import LabelDiv from './widgets/LabelDiv.vue';
import EditableTextWithSuggestions from './widgets/EditableTextWithSuggestions.vue';

const owner = defineModel<string | null>({ default: null });
const userStore = useUserStore();

async function selfAssign() {
  isLoading.value = true;
  return userStore.updateJiraUsername().then(() => {
    if (userStore.jiraUsername !== '') {
      owner.value = userStore.jiraUsername;
    }
  }).finally(() => {
    isLoading.value = false;
  });
}

async function handleClick(fn: (arg?: any) => any) {
  await selfAssign();
  nextTick(fn);
}

const isAssignedToMe = computed(() =>
  owner.value === userStore.jiraUsername && userStore.jiraUsername !== ''
);

const isLoading = ref(false);
const results = ref<ZodJiraUserPickerType[]>([]);

const cacheJiraUsers = useMemoize(async (query: string) => {
  if (!query) {
    return [];
  }
  isLoading.value = true;
  const users = await searchJiraUsers(query)
    .catch(createCatchHandler('Failed to search jira users', false));
  isLoading.value = false;
  return users?.data?.users ?? [];
});


const onQueryChange = async (query: string) => {
  const users = await cacheJiraUsers(query);
  results.value = users ?? [];
};

const handleSuggestionClick = (fn: (args?: any) => void, user: ZodJiraUserPickerType) => {
  owner.value = toValue(user.name);
  results.value = [];
  nextTick(fn);
};
</script>

<template>
  <LabelDiv label="Owner" :loading="isLoading">
    <EditableTextWithSuggestions v-model="owner" class="col-12" @update:query="onQueryChange">
      <template v-if="!isAssignedToMe" #buttons-out-of-editing-mode="{ onBlur }">
        <button
          type="button"
          class="btn btn-primary osim-self-assign"
          :disabled="isLoading"
          @click.prevent.stop="handleClick(onBlur)"
        >
          Self Assign
        </button>
      </template>
      <template v-if="!isAssignedToMe" #buttons-in-editing-mode="{ onBlur }">
        <button
          type="button"
          class="btn btn-primary osim-self-assign"
          :disabled="isLoading"
          @click.prevent.stop="handleClick(onBlur)"
        >
          Self Assign
        </button>
      </template>
      <template v-if="results.length > 0" #suggestions="{ onBlur }">
        <div
          v-for="user in results"
          :key="user.name"
          class="item"
          @click.prevent.stop="handleSuggestionClick(onBlur, user)"
        >
          <span v-html="sanitize(user.html)" />
        </div>
      </template>
    </EditableTextWithSuggestions>
  </LabelDiv>
</template>
