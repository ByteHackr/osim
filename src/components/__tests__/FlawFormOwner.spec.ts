import { describe, it, expect, vi } from 'vitest';
import { createTestingPinia } from '@pinia/testing';
import FlawFormOwner from '../FlawFormOwner.vue';
import { flushPromises, mount, VueWrapper } from '@vue/test-utils';
import { useUserStore } from '@/stores/UserStore';
import { searchJiraUsers } from '@/services/JiraService';

const pinia = createTestingPinia();
const userStore = useUserStore();

vi.mock('@vueuse/core', async (originalImport) => ({
  ... (await originalImport<typeof import('@vueuse/core')>()),
  useLocalStorage: vi.fn((key: string) => {
    return {
      UserStore: {
        value: {
          refresh: 'mocked_refresh_token',
          env: 'mocked_env',
          whoami: {
            email: 'test@example.com',
          },
          jiraUsername: 'skynet',
        },
      },
    }[key];
  }),
  useStorage: vi.fn((key: string, defaults) => {
    return {
      'OSIM::API-KEYS': {
        value: defaults || {
          bugzillaApiKey: '',
          jiraApiKey: '',
          showNotifications: false,
        },
      },
    }[key];
  }),

}));

vi.mock('@/services/JiraService', () => ({
  searchJiraUsers: vi.fn(() => Promise.resolve([])),
}));

vi.mock('jwt-decode', () => ({
  default: vi.fn(() => ({
    sub: '1234567890',
    name: 'Test User',
    exp: Math.floor(Date.now() / 1000) + 60 * 60 * 24 * 365,
  })),
}));


describe('Owner field', () => {
  let subject: VueWrapper<InstanceType<typeof FlawFormOwner>>;
  beforeEach(() => {
    vi.useFakeTimers();
    userStore.updateJiraUsername = vi.fn(() => Promise.resolve());
    subject = mount(FlawFormOwner, {
      global: {
        plugins: [pinia],
      },
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
    vi.useRealTimers();
  });

  it('should render the owner field', () => {
    expect(subject.exists()).toBe(true);
  });

  it('assigns the test user when button is clicked', async () => {
    await subject.find('button.osim-self-assign').trigger('click');
    await subject.vm.$nextTick();
    expect(subject.text()).toContain('skynet');
  });

  it('should call "searchJiraUsers" when input changes', async () => {
    const input = subject.find('input');
    await input.setValue('test');

    await vi.runAllTimers();

    expect(searchJiraUsers).toHaveBeenCalled();
  });

  it('should show results when input changes', async () => {
    vi.mocked(searchJiraUsers, { partial: true }).mockResolvedValueOnce({
      data: {
        users:
          [{ name: 'test', displayName: 'Test User', html: 'Test User' }]
      }
    });
    const input = subject.find('input');
    await input.setValue('test');

    await vi.runAllTimers();
    await flushPromises();

    expect(subject.text()).toContain('Test User');
  });
});
