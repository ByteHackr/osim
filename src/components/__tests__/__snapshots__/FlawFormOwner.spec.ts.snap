// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`owner field > assigns the test user when button is clicked 1`] = `
"<div data-v-0e2ae48e="" class="osim-input ps-3 mb-2 mb-2">
  <div data-v-0e2ae48e="" class="row"><span data-v-0e2ae48e="" class="form-label col-3 position-relative"><!--v-if--> Owner <!--attrs: {{ $attrs }}--></span>
    <div data-v-0e2ae48e="" class="col-9">
      <!-- for invalid-tooltip positioning -->
      <div class="position-relative col-9 osim-editable-field osim-text col-12">
        <!--<Transition name="flash-bg" :duration="2000">-->
        <transition-stub name="flash-bg" appear="false" persisted="true" css="true">
          <div class="osim-editable-text" tabindex="0"><span class="osim-editable-text-value form-control">skynet</span>
            <!--if a button is inside a label, clicking the label clicks the button?--><button type="button" class="osim-editable-text-pen input-group-text" tabindex="-1"><i class="bi bi-pencil"></i></button>
          </div>
        </transition-stub>
        <div class="input-group position-relative" tabindex="-1" style="display: none;">
          <!--v-if--><input class="form-control" type="text"><button type="button" class="input-group-text" tabindex="-1"><i class="bi bi-check"></i></button><button type="button" class="input-group-text" tabindex="-1"><i class="bi bi-x"></i></button>
        </div>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`owner field > should render the owner field 1`] = `
"<div data-v-0e2ae48e="" class="osim-input ps-3 mb-2 mb-2">
  <div data-v-0e2ae48e="" class="row"><span data-v-0e2ae48e="" class="form-label col-3 position-relative"><!--v-if--> Owner <!--attrs: {{ $attrs }}--></span>
    <div data-v-0e2ae48e="" class="col-9">
      <!-- for invalid-tooltip positioning -->
      <div class="position-relative col-9 osim-editable-field osim-text col-12">
        <!--<Transition name="flash-bg" :duration="2000">-->
        <transition-stub name="flash-bg" appear="false" persisted="true" css="true">
          <div class="osim-editable-text" tabindex="0"><span class="osim-editable-text-value form-control"></span>
            <!--if a button is inside a label, clicking the label clicks the button?--><button type="button" class="osim-editable-text-pen input-group-text" tabindex="-1"><i class="bi bi-pencil"></i></button><button type="button" class="btn btn-primary osim-self-assign"> Self Assign </button>
          </div>
        </transition-stub>
        <div class="input-group position-relative" tabindex="-1" style="display: none;">
          <!--v-if--><input class="form-control" type="text"><button type="button" class="input-group-text" tabindex="-1"><i class="bi bi-check"></i></button><button type="button" class="input-group-text" tabindex="-1"><i class="bi bi-x"></i></button><button type="button" class="btn btn-primary osim-self-assign"> Self Assign </button>
        </div>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;
